### PR TITLE
Branch list tidy up

### PR DESCRIPTION
**Pull request by qtzar**
Sorts the list of branches in a multi-branch site and moves the default branch to the top of the list.

### CODE REVIEW

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/GenerateSiteCommand.kt [hunk 1]
```diff
@@ -82,7 +82,7 @@ class GenerateSiteCommand : Subcommand(
             refreshLocalClone()
         }
 
-        val branchNames = if (allBranches)
+        val branchNames: List<String> = if (allBranches)
```
Comment (jp7677) [inline src/main/kotlin/nl/avisi/structurizr/site/generatr/GenerateSiteCommand.kt:85]: This change is not needed.

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/GenerateSiteCommand.kt [hunk 2]
```diff
@@ -109,7 +109,12 @@ class GenerateSiteCommand : Subcommand(
             throw Exception("$defaultBranch does not contain a valid structurizr workspace. Site generation halted.")
         }
 
-        branchesToGenerate.forEach { branch ->
+        val sortedBranchNames = branchesToGenerate.toMutableList()
```
Comment (jp7677) [inline src/main/kotlin/nl/avisi/structurizr/site/generatr/GenerateSiteCommand.kt:112]: We can do this more elegantly. My proposal:

Introduce a custom comparator for our needs on the top level of this file:

```kotlin
fun branchComparator(defaultBranch: String) = Comparator<String> { a, b ->
    if (a == defaultBranch) -1
    else if (b == defaultBranch) 1
    else a.compareTo(b)
}
```

The test for that one would look like:

```kotlin
class BranchComparatorTest {
    @TestFactory
    fun `default branch first then by string`() = listOf(
        listOf("a", "b", "main"),
        listOf("a", "main", "b"),
        listOf("b", "a", "main"),
        listOf("b", "main", "a"),
        listOf("main", "a", "b"),
        listOf("main", "b", "a")
    ).map { branches ->
        DynamicTest.dynamicTest(branches.toString()) {
            assertThat(branches.sortedWith(branchComparator("main")))
                .containsExactly("main", "a", "b")
        }
    }
}
```

Having this in place we can just add the following to line 104 to have `branchesToGenerate` already sorted on initialization:


```kotlin
}.sortedWith(branchComparator(defaultBranch))
```

**Review by jp7677 — CHANGES_REQUESTED** [review]
See comments.

**Comment by qtzar** [comment]
@jp7677 It feels weird having the branch comparator function sitting outside the class. Probably because I'm just not that used to Kotlin :)

**Review by jp7677 — APPROVED** [review]

**Comment by jp7677** [comment]
> @jp7677 It feels weird having the branch comparator function sitting outside the class. Probably because I'm just not that used to Kotlin :)

I guess both, outside the class or as companion object within the class,  would be fine as long as writing tests is still easy. I opted for outside since you could say that sorting with that order is a kind-of general thing. I guess linters would suggest to put it into a separate file with the name `BranchComparator.kt`.

**Pull request merged**